### PR TITLE
Log detail name for optimizers of RowExpressionRewriteRuleSet and StatsRecordingPlanOptimizer

### PR DESCRIPTION
**Pull request by feilong-liu**
## Description
Currently optimizer log will record the class name of the optimizer, however this can be a problem for row expression optimizers, which extends RowExpressionRewriteRuleSet. All such optimizers will have the same name in the log, like "FilterRowExpressionRewrite", "ValuesRowExpressionRewrite" etc. These give no meaningful information on which optimizers are triggered.

This PR will log more verbose information for these optimizers, by including the rewriter class name.
For example instead of "ProjectRowExpressionRewrite" we will log RemoveMapCastRule$Rewriter:ProjectRowExpressionRewrite" now.

Similar for StatsRecordingPlanOptimizer, it's a wrapping optimizer, and it's the delegate optimizer within it doing the work. After the change, it will show as "StatsRecordingPlanOptimizer:HistoricalStatisticsEquivalentPlanMarkingOptimizer" instead of "StatsRecordingPlanOptimizer".

## Motivation and Context
To make optimizer log more meaningful.

## Impact
Improve optimizer log

## Test Plan
Test locally end to end

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Improve logging for RowExpressionRewriteRuleSet and StatsRecordingPlanOptimizer optimizers to include more information :pr:`22765`
```

### CODE REVIEW

**Comment by jaystarshot** [comment]
This also happens for StatsRecordingPlanOptimizer, wondering if we could fix that too

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java [hunk 1]
```diff
@@ -210,6 +211,45 @@ private boolean exploreNode(int group, Context context, Matcher matcher)
...
+        else if (rule instanceof RowExpressionRewriteRuleSet.ValuesRowExpressionRewrite) {
+            ruleName = ((RowExpressionRewriteRuleSet.ValuesRowExpressionRewrite) rule).getOptimizerNameForLog();
+        }
+        else if (rule instanceof RowExpressionRewriteRuleSet.AggregationRowExpressionRewrite) {
+            ruleName = ((RowExpressionRewriteRuleSet.AggregationRowExpressionRewrite) rule).getOptimizerNameForLog();
+        }
+        else if (rule instanceof RowExpressionRewriteRuleSet.TableFinishRowExpressionRewrite) {
+            ruleName = ((RowExpressionRewriteRuleSet.TableFinishRowExpressionRewrite) rule).getOptimizerNameForLog();
+        }
+        else if (rule instanceof RowExpressionRewriteRuleSet.TableWriterRowExpressionRewrite) {
```
Comment (jaystarshot) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:247]: On adding a new rule you will have to extend this, wondering if there was a better way

Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:247]: Yeah, I was trying to find a better way to achieve this but cannot think of one. Open to suggestions if there are better ways.

**Comment by feilong-liu** [comment]
> This also happens for `StatsRecordingPlanOptimizer` i.e optimizers which extend that show up as StatsRecordingPlanOptimizer, wondering if we could fix that too but that would be a different PR

This is not a big change, added in this PR too.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java [hunk 2]
```diff
@@ -210,6 +211,45 @@ private boolean exploreNode(int group, Context context, Matcher matcher)
...
+        if (rule instanceof RowExpressionRewriteRuleSet.ProjectRowExpressionRewrite) {
+            ruleName = ((RowExpressionRewriteRuleSet.ProjectRowExpressionRewrite) rule).getOptimizerNameForLog();
+        }
+        else if (rule instanceof RowExpressionRewriteRuleSet.SpatialJoinRowExpressionRewrite) {
+            ruleName = ((RowExpressionRewriteRuleSet.SpatialJoinRowExpressionRewrite) rule).getOptimizerNameForLog();
+        }
+        else if (rule instanceof RowExpressionRewriteRuleSet.JoinRowExpressionRewrite) {
+            ruleName = ((RowExpressionRewriteRuleSet.JoinRowExpressionRewrite) rule).getOptimizerNameForLog();
+        }
+        else if (rule instanceof RowExpressionRewriteRuleSet.JoinRowExpressionRewrite) {
```
Comment (jaystarshot) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:226]: This else is repeated with above

**Review by jaystarshot — DISMISSED** [review]
Don't see an immediate solution to better log rules, maybe rules need to have a field with info about the optimizer which is using but this can be refactored if needed in future.

**Comment by jaystarshot** [comment]
@ClarenceThreepwood  Do you have any other ideas?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java (hunk 1 above)
Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:244]: Would the following patch work for you @feilong-liu ? I think you can just extend the rule interface to add a method which does this, and extend all of the rules using the class with the method instead, and check if it is an instance of the abstract class e.g.

```java
    public abstract class RowExpressionRewriteRule<T> implements Rule<T> {
        public String getOptimizerNameForLog() {
            String rewriterName = rewriter.getClass().getName();
            return format("%s:%s", rewriterName.substring(rewriterName.lastIndexOf('.') + 1), this.getClass().getSimpleName());
        }
    }

// rule definition
public final class ProjectRowExpressionRewrite
            extends RowExpressionRewriteRule<ProjectNode>
...
// name check 
        if (rule instanceof RowExpressionRewriteRuleSet.RowExpressionRewriteRule) {
            ruleName = ((RowExpressionRewriteRuleSet.RowExpressionRewriteRule) rule).getOptimizerNameForLog();
        }
```

[simplify-rulename.patch](https://github.com/prestodb/presto/files/15343575/simplify-rulename.patch)

Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java]: Good suggestion. Applied the change.

**Review by jaystarshot — APPROVED** [review]

**Pull request merged**